### PR TITLE
Weed update

### PR DESCRIPTION
**Pull request by guol-fnst**
# What problem are we solving?
update weed from github
[#3161 ](https://github.com/chrislusf/seaweedfs/issues/3161)

# How are we solving the problem?
On linux, "weed update" gets the latest version from github and save it to a new file or replace  the old weed.
On window, "weed update" gets the latest version from github and save it to a new file .
copied from "https://github.com/restic/restic/tree/master/internal/selfupdate"

# Checks
- [ no] I have added unit tests if possible.
- [ yes] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

**Comment by chrislusf** [comment]
I tried it, but I can not tell where is the output binary.
```
seaweedfs/weed () - (pr/3184) > weed update
I0615 00:48:29 97890 update.go:94] writing weed to weed
I0615 00:48:29 97890 update.go:115] weed is up to date
I0615 00:48:29 97890 update.go:102] successfully updated weed to version 3.11
seaweedfs/weed () - (pr/3184) > ls -al weed
ls: weed: No such file or directory
```

**Comment by guol-fnst** [comment]
> I tried it, but I can not tell where is the output binary.
> 
> ```
> seaweedfs/weed () - (pr/3184) > weed update
> I0615 00:48:29 97890 update.go:94] writing weed to weed
> I0615 00:48:29 97890 update.go:115] weed is up to date
> I0615 00:48:29 97890 update.go:102] successfully updated weed to version 3.11
> seaweedfs/weed () - (pr/3184) > ls -al weed
> ls: weed: No such file or directory
> ```

This log "I0615 00:48:29 97890 update.go:102] successfully updated weed to version 3.11" maybe misleading.
When the current version is up to date ,there's no need to download weed. 
I add some fix to remove this log when weed is up to date.

Please try again

**Comment by guol-fnst** [comment]
Here is the output
```
[root@localhost weed]# ./weed update
I0615 16:02:07 32418 update.go:116] latest version is 3.11
I0615 16:02:07 32418 update.go:266] download linux_amd64.tar.gz.md5
I0615 16:02:08 32418 update.go:266] download linux_amd64.tar.gz
I0615 16:02:17 32418 update.go:350] saved 78129990 bytes in weed
I0615 16:02:17 32418 update.go:153] successfully updated weed to version 3.11
[root@localhost weed]# ./weed version
version 30GB 3.11 d4ef06cdcf320f8b8b17279586e0738894869eff linux amd64

[root@localhost weed]# ./weed update
I0615 16:02:51 32659 update.go:112] weed is up to date

```

**Comment by guol-fnst** [comment]
@chrislusf  I find that I didn't check whether it's a "FULL" version. I'll add this feature later.

**Comment by guol-fnst** [comment]
@chrislusf 
I have added checks for full version weed.
Now "weed update" can update largedisk/full version also.

**Comment by guol-fnst** [comment]
> Could you please add an option to force download to a specific version?

Sure. I'll add the feature tomorrow.

**Comment by chrislusf** [comment]
Also, add an option to specify the output directory

**Comment by guol-fnst** [comment]
> Also, add an option to specify the output directory

Now "weed update ' supports the following pattern:
1)./weed update -version=3.1 -output=/tmp/weed #save version 3.1 to /tmp/weed
2)./weed update -version=3.1 -output=weed_new #save version 3.1 to weed_new
3)./weed update -version=3.1  #save version 3.1 to weed
4)./weed update --output=/tmp/weed #save latest version to /tmp/weed
5)./weed update #save the latest version to weed

And if the current version equals the specific version, weed update will exit.

**Comment by chrislusf** [comment]
* can output be a directory? The output file name can be derived from current binary name. 
* The path can actually be the derived from the current binary also if it is not on windows.

File: weed/command/update.go [hunk 1]
```diff
@@ -77,26 +81,33 @@ var cmdUpdate = &Command{
...
+		*updateOpt.dir = filepath.Dir(path)
+	}
+
+	if *updateOpt.name == "" {
+		*updateOpt.name = name
+	}
+
+	if runtime.GOOS == "windows" {
+		if *updateOpt.name == name || *updateOpt.name == "" {
+			glog.Fatalf("On windows, name of the new weed shouldn't be same to the orignial name.")
```
Comment (chrislusf) [inline weed/command/update.go:104]: Only when both dir and name are the same.

File: weed/command/update.go [hunk 2]
```diff
@@ -77,26 +81,33 @@ var cmdUpdate = &Command{
...
 			return false
 		}
 	}
 
-	_, err = downloadRelease(context.Background(), *updateOpt.Output, *updateOpt.Version)
+	target := *updateOpt.dir + "/" + *updateOpt.name
```
Comment (chrislusf) [inline weed/command/update.go]: This does not work on windows.

File: weed/command/update.go (hunk 1 above)
Comment (guol-fnst) [reply weed/command/update.go:104]: done
Tested on windows.

File: weed/command/update.go (hunk 2 above)
Comment (guol-fnst) [reply weed/command/update.go]: done
Tested on windows.

File: weed/command/update.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+	path, _ := os.Executable()
+	_, name := filepath.Split(path)
+	updateOpt.dir = cmdUpdate.Flag.String("dir", filepath.Dir(path), "directory to save new weed.")
+	updateOpt.name = cmdUpdate.Flag.String("name", name, "name of new weed. On windows, name shouldn't be same to the orignial name.")
+	updateOpt.Version = cmdUpdate.Flag.String("version", "0", "specific version of weed you want to download. If not specified, get the latest version.")
+	cmdUpdate.Run = runUpdate
+}
+
+var cmdUpdate = &Command{
+	UsageLine: "update [-output=weed]",
```
Comment (chrislusf) [inline weed/command/update.go]: usage needs update

Comment (guol-fnst) [reply weed/command/update.go]: done

**Pull request merged**